In [1]:
!pip install datasets

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 365 kB 4.7 MB/s 
     |████████████████████████████████| 141 kB 54.0 MB/s 
     |████████████████████████████████| 115 kB 54.3 MB/s 
     |████████████████████████████████| 212 kB 62.1 MB/s 
     |████████████████████████████████| 101 kB 13.7 MB/s 
     |████████████████████████████████| 596 kB 62.4 MB/s 
     |████████████████████████████████| 127 kB 73.7 MB/s 
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.24.3
    Uninstalling urllib3-1.24.3:
      Successfully uninstalled urllib3-1.24.3
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [2]:
!git clone https://github.com/drehero/geneval

Cloning into 'geneval'...
remote: Enumerating objects: 264, done.
remote: Counting objects: 100% (59/59), done.
remote: Compressing objects: 100% (44/44), done.
remote: Total 264 (delta 15), reused 53 (delta 11), pack-reused 205
Receiving objects: 100% (264/264), 11.11 MiB | 9.57 MiB/s, done.
Resolving deltas: 100% (89/89), done.


In [3]:
import datasets
import numpy as np

from geneval.geneval.data.wmt import WMT18

In [4]:
# kendall score
# https://github.com/Tiiiger/bert_score/blob/master/reproduce/get_wmt18_seg_results.py#L53

def kendall_score(scores_better, scores_worse):
    total = len(scores_better)
    correct = np.sum(np.array(scores_better) > np.array(scores_worse))
    incorrect = total - correct
    return (correct - incorrect)/total

In [5]:
pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 4.7 MB 4.0 MB/s 
     |████████████████████████████████| 6.6 MB 38.1 MB/s 


### FrugalScore

https://huggingface.co/spaces/evaluate-metric/frugalscore

In [ ]:
frugal_score = datasets.load_metric("frugalscore")

In [ ]:
frugal_score_results = {}
for lang_pair in ["cs-en", "de-en", "et-en", "fi-en", "ru-en", "tr-en", "zh-en"]:
    # load data
    wmt = WMT18(lang_pair, root="/tmp")
    # compute scores
    scores_better = frugal_score.compute(predictions = wmt.translations_better,
        references = wmt.references
    )
    scores_worse = frugal_score.compute(
        predictions = wmt.translations_worse,
        references = wmt.references
    )
    ks = kendall_score(scores_better["scores"], scores_worse["scores"])
    frugal_score_results[lang_pair] = ks

  0%|          | 0/6 [00:00<?, ?ba/s]

  0%|          | 0/6 [00:00<?, ?ba/s]

  0%|          | 0/78 [00:00<?, ?ba/s]

  0%|          | 0/78 [00:00<?, ?ba/s]

  0%|          | 0/57 [00:00<?, ?ba/s]

  0%|          | 0/57 [00:00<?, ?ba/s]

  0%|          | 0/16 [00:00<?, ?ba/s]

  0%|          | 0/16 [00:00<?, ?ba/s]

  0%|          | 0/11 [00:00<?, ?ba/s]

  0%|          | 0/11 [00:00<?, ?ba/s]

  0%|          | 0/9 [00:00<?, ?ba/s]

  0%|          | 0/9 [00:00<?, ?ba/s]

  0%|          | 0/34 [00:00<?, ?ba/s]

  0%|          | 0/34 [00:00<?, ?ba/s]

In [ ]:
frugal_score_results

{'cs-en': 0.30293542074363994,
 'de-en': 0.4720026731439,
 'et-en': 0.3388339415736676,
 'fi-en': 0.22098670756646216,
 'ru-en': 0.28565936178392926,
 'tr-en': 0.22205278592375366,
 'zh-en': 0.20016788080462872}

Frugal score: 

{'cs-en': 0.30293542074363994,
 'de-en': 0.4720026731439,
 'et-en': 0.3388339415736676,
 'fi-en': 0.22098670756646216,
 'ru-en': 0.28565936178392926,
 'tr-en': 0.22205278592375366,
 'zh-en': 0.20016788080462872}

### BLEURT

https://huggingface.co/spaces/evaluate-metric/bleurt

In [6]:
pip install git+https://github.com/google-research/bleurt.git

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/google-research/bleurt.git to /tmp/pip-req-build-xtfomjha
  Running command git clone -q https://github.com/google-research/bleurt.git /tmp/pip-req-build-xtfomjha
     |████████████████████████████████| 352 kB 4.8 MB/s 
     |████████████████████████████████| 1.2 MB 46.0 MB/s 
  Created wheel for BLEURT: filename=BLEURT-0.0.2-py3-none-any.whl size=16456761 sha256=d9a46b406fdb00f788b53d542aeec1229d6ec10417503ea8c5888e63ed0761eb
  Stored in directory: /tmp/pip-ephem-wheel-cache-qtadseio/wheels/e2/2d/ea/b7a8b2424d2908d2a79d73ce8217d5ac4bd97ed3f47160a7f5
Successfully built BLEURT


In [7]:
bleurt = datasets.load_metric("bleurt")

Using default BLEURT-Base checkpoint for sequence maximum length 128. You can use a bigger model for better results with e.g.: datasets.load_metric('bleurt', 'bleurt-large-512').


INFO:tensorflow:Reading checkpoint /root/.cache/huggingface/metrics/bleurt/default/downloads/extracted/887f2dc36c17f53c287f696681b8f7c947278407c1cf9f226662e16c8c0dc417/bleurt-base-128.
INFO:tensorflow:Config file found, reading.
INFO:tensorflow:Will load checkpoint bert_custom
INFO:tensorflow:Loads full paths and checks that files exists.
INFO:tensorflow:... name:bert_custom
INFO:tensorflow:... vocab_file:vocab.txt
INFO:tensorflow:... bert_config_file:bert_config.json
INFO:tensorflow:... do_lower_case:True
INFO:tensorflow:... max_seq_length:128
INFO:tensorflow:Creating BLEURT scorer.
INFO:tensorflow:Creating WordPiece tokenizer.
INFO:tensorflow:WordPiece tokenizer instantiated.
INFO:tensorflow:Creating Eager Mode predictor.
INFO:tensorflow:Loading model.
INFO:tensorflow:BLEURT initialized.


In [8]:
predictions = ["hello there", "general kenobi"]
references = ["hello there", "general kenobi"]
results = bleurt.compute(predictions=predictions, references=references)

print(results)

{'scores': [1.0295498371124268, 1.0445427894592285]}


In [9]:
bleurt_results = {}
for lang_pair in ["cs-en", "de-en", "et-en", "fi-en", "ru-en", "tr-en", "zh-en"]:
    # load data
    wmt = WMT18(lang_pair, root="/tmp")
    # compute scores
    scores_better = bleurt.compute(
        predictions = [wmt.translations_better],
        references = [wmt.references]
    )
    scores_worse = bleurt.compute(
        predictions = [wmt.translations_worse],
        references = [wmt.references]
    )
    ks = kendall_score(scores_better["scores"], scores_worse["scores"])
    bleurt_results[lang_pair] = ks

In [10]:
bleurt_results

{'cs-en': 1.0,
 'de-en': 1.0,
 'et-en': 1.0,
 'fi-en': 1.0,
 'ru-en': 1.0,
 'tr-en': 1.0,
 'zh-en': 1.0}

BLEURT:

{'cs-en': 1.0,
 'de-en': 1.0,
 'et-en': 1.0,
 'fi-en': 1.0,
 'ru-en': 1.0,
 'tr-en': 1.0,
 'zh-en': 1.0}